src/protocols/bdx/BdxTransferSession.cpp

```diff
@@ -0,0 +1,934 @@
+
+    err = payloadHeader.Decode(msg->Start(), msg->DataLength(), &headerSize);
+    SuccessOrExit(err);
+
+    msg->ConsumeHead(headerSize);
+
+    if (payloadHeader.GetProtocolID() == Protocols::kProtocol_BDX)
+    {
+        err = HandleBdxMessage(payloadHeader, std::move(msg));
+        SuccessOrExit(err);
+
+        mTimeoutStartTimeMs = curTimeMs;
+    }
+    else if (payloadHeader.GetProtocolID() == Protocols::kProtocol_Protocol_Common &&
+             payloadHeader.GetMessageType() == static_cast<uint8_t>(Protocols::Common::MsgType::StatusReport))
+    {
+        err = HandleStatusReportMessage(payloadHeader, std::move(msg));
+        SuccessOrExit(err);
+    }
+    else
+    {
+        err = CHIP_ERROR_INVALID_MESSAGE_TYPE;
+    }
+
+exit:
+    return err;
+}
+
+// Return CHIP_ERROR only if there was a problem decoding the message. Otherwise, call PrepareStatusReport().
+CHIP_ERROR TransferSession::HandleBdxMessage(PayloadHeader & header, System::PacketBufferHandle msg)
+{
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    MessageType msgType = static_cast<MessageType>(header.GetMessageType());
+
+    VerifyOrExit(!msg.IsNull(), err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(mPendingOutput == kNone, err = CHIP_ERROR_INCORRECT_STATE);
+
+    switch (msgType)
+    {
+    case kBdxMsg_SendInit:
+    case kBdxMsg_ReceiveInit:
+        HandleTransferInit(msgType, std::move(msg));
+        break;
+    case kBdxMsg_SendAccept:
+        HandleSendAccept(std::move(msg));
+        break;
+    case kBdxMsg_ReceiveAccept:
+        HandleReceiveAccept(std::move(msg));
+        break;
+    case kBdxMsg_BlockQuery:
+        HandleBlockQuery(std::move(msg));
+        break;
+    case kBdxMsg_Block:
+        HandleBlock(std::move(msg));
+        break;
+    case kBdxMsg_BlockEOF:
+        HandleBlockEOF(std::move(msg));
+        break;
+    case kBdxMsg_BlockAck:
+        HandleBlockAck(std::move(msg));
+        break;
+    case kBdxMsg_BlockAckEOF:
+        HandleBlockAckEOF(std::move(msg));
+        break;
+    default:
+        err = CHIP_ERROR_INVALID_MESSAGE_TYPE;
+        break;
+    }
+
+exit:
+    return err;
+}
+
+/**
+ * @brief
+ *   Parse a StatusReport message and prepare to emit an OutputEvent with the message data.
+ *
+ *   NOTE: BDX does not currently expect to ever use a "Success" general code, so it will be treated as an error along with any
+ *         other code.
+ */
+CHIP_ERROR TransferSession::HandleStatusReportMessage(PayloadHeader & header, System::PacketBufferHandle msg)
+{
+    VerifyOrReturnError(!msg.IsNull(), CHIP_ERROR_INVALID_ARGUMENT);
+
+    mState            = kErrorState;
+    mAwaitingResponse = false;
+
+    uint16_t generalCode  = 0;
+    uint32_t protocolId   = 0;
+    uint16_t protocolCode = 0;
+    Encoding::LittleEndian::Reader reader(msg->Start(), msg->DataLength());
+    ReturnErrorOnFailure(reader.Read16(&generalCode).Read32(&protocolId).Read16(&protocolCode).StatusCode());
```

Comment:
Shouldn't we have a common function to parse status reports?

Reply:
I agree, I just don't think anyone's implemented this yet. I think any common implementation would probably belong in a separate PR